src/main/java/net/neoforged/neoforge/client/model/generators/template/ElementBuilder.java

```diff
@@ -0,0 +1,257 @@
+    private Vector3f from = new Vector3f();
+    private Vector3f to = new Vector3f(16, 16, 16);
+    private final Map<Direction, FaceBuilder> faces = new LinkedHashMap<>();
+    @Nullable
+    private RotationBuilder rotation;
+    private boolean shade = true;
+    private int lightEmission = 0;
+    private int color = 0xFFFFFFFF;
+    private int blockLight = 0;
+    private int skyLight = 0;
+    private boolean hasAmbientOcclusion = true;
+
+    private static void validateCoordinate(float coord, char name) {
+        Preconditions.checkArgument(!(coord < -16.0F) && !(coord > 32.0F), "Position " + name + " out of range, must be within [-16, 32]. Found: %d", coord);
+    }
+
+    private static void validatePosition(Vector3f pos) {
+        validateCoordinate(pos.x(), 'x');
+        validateCoordinate(pos.y(), 'y');
+        validateCoordinate(pos.z(), 'z');
+    }
+
+    /**
+     * Set the "from" position for this element.
+     *
+     * @param x x-position for this vector
+     * @param y y-position for this vector
+     * @param z z-position for this vector
+     * @return this builder
+     * @throws IllegalArgumentException if the vector is out of bounds (any
+     *                                  coordinate not between -16 and 32,
+     *                                  inclusive)
+     */
+    public ElementBuilder from(float x, float y, float z) {
+        this.from = new Vector3f(x, y, z);
+        validatePosition(this.from);
+        return this;
+    }
+
+    /**
+     * Set the "to" position for this element.
+     *
+     * @param x x-position for this vector
+     * @param y y-position for this vector
+     * @param z z-position for this vector
+     * @return this builder
+     * @throws IllegalArgumentException if the vector is out of bounds (any
+     *                                  coordinate not between -16 and 32,
+     *                                  inclusive)
+     */
+    public ElementBuilder to(float x, float y, float z) {
+        this.to = new Vector3f(x, y, z);
+        validatePosition(this.to);
+        return this;
+    }
+
+    /**
+     * Return or create the face builder for the given direction.
+     *
+     * @param dir the direction
+     * @return the face builder for the given direction
+     * @throws NullPointerException if {@code dir} is {@code null}
+     */
+    public ElementBuilder face(Direction dir, Consumer<FaceBuilder> action) {
+        Preconditions.checkNotNull(dir, "Direction must not be null");
+        var builder = faces.computeIfAbsent(dir, $ -> new FaceBuilder());
+        action.accept(builder);
+        return this;
+    }
+
+    public ElementBuilder rotation(Consumer<RotationBuilder> action) {
+        if (this.rotation == null) {
+            this.rotation = new RotationBuilder();
+        }
+        action.accept(rotation);
+        return this;
+    }
+
+    public ElementBuilder shade(boolean shade) {
+        this.shade = shade;
+        return this;
+    }
```

Comment:
These should have short doc comments, particularly shade as that seems to be something that regularly confuses people in relation to AO.